hypothesis-python/src/hypothesis/internal/conjecture/engine.py

```diff
@@ -992,12 +1028,36 @@ def add(self, data):
             self.best_status = data.status
             self.reset()
 
-        self.fresh_examples.append(data)
+        updated = False
+        self.check()
+        for key, score in data.target_observations.items():
+            if key not in self.label_scores or score > self.label_scores[key]:
+                self.label_scores[key] = score
+                self.best_for_label_fresh[key] = data
+                self.best_for_label_used.pop(key, None)
+                updated = True
+        self.check()
+
+        if not updated:
+            self.fresh_examples.append(data)
         if len(self) > self.pool_size:
             pop_random(self.random, self.used_examples or self.fresh_examples)
             assert self.pool_size == len(self)
 
     def select(self):
+        # If there is a target that is the current best known for a particular
+        # example and we have not mutated it before, try that first.
+        best_fresh = self.best_for_label_fresh
+        if best_fresh:
+            value = best_fresh[self.random.choice(list(best_fresh))]
+            for key in [k for k, data in best_fresh.items() if data == value]:
```

Comment:
Any reason why this a list-comprehension instead of a generator?

Reply:
Yep: it's an error to change the size of a dictionary while iterating over it, so we can't take the lazy way out (😉).

Reply:
Ah! The nefarious pop! I see it now. Thanks. (This looks like a really cool feature, btw! I hope to find more time to dive into it)

Reply:
Be warned: amazing concept, but the implementation is more "minimum" than "viable" - I really prioritised getting the API right for the long term and getting something to evaluate over doing well on evaluations. But 'make it work, make it right, make it fast'... eventually!